go-controller/pkg/ovn/address_set_sync.go

```diff
@@ -0,0 +1,358 @@
+		if err != nil {
+			err = fmt.Errorf("failed to get acl delete from logical switch ops: %v", err)
+			return
+		}
+	}
+	if len(qoses) > 0 {
+		ops, err = libovsdbops.RemoveQoSsFromLogicalSwitchesWithPredicateOps(syncer.nbClient, ops,
+			func(item *nbdb.LogicalSwitch) bool { return true }, qoses...)
+		if err != nil {
+			err = fmt.Errorf("failed to get qos delete from logical switch ops: %v", err)
+			return
+		}
+		ops, err = libovsdbops.DeleteQoSesOps(syncer.nbClient, ops, qoses...)
+		if err != nil {
+			err = fmt.Errorf("failed to get qos delete ops: %v", err)
+			return
+		}
+	}
+	if len(lrps) > 0 {
+		ops, err = libovsdbops.DeleteLogicalRouterPoliciesOps(syncer.nbClient, ops, ovntypes.OVNClusterRouter, lrps...)
+		if err != nil {
+			err = fmt.Errorf("failed to get logical router policy delete ops: %v", err)
+			return
+		}
+	}
+	ops, err = libovsdbops.DeleteAddressSetsOps(syncer.nbClient, ops, as)
+	return
+}
+
+// updateOpsForAS adds db ops to update address set and objects that reference it
+func (syncer *addressSetsSyncer) updateOpsForAS(as *nbdb.AddressSet) error {
+	oldName, ipSuffix := truncateSuffixFromAddressSet(as.ExternalIDs["name"])
+	// oldName may be empty if as doesn't have ExternalID set
+	acls, qoses, lrps, dbIndex, err := GetReferencingObjsAndNewDbIndex(syncer.nbClient, as.Name, oldName)
+	if err != nil {
+		return fmt.Errorf("failed to get new dbIndex for address set %s: %v", oldName, err)
+	}
+
+	if oldName == "" {
+		klog.Infof("Found stale address set %s with empty ExternalIDs[\"name\"], deleting", as.Name)
+		// wrong format, must be stale, delete
+		ops, err := syncer.getDeleteASOps(acls, qoses, lrps, as)
+		if err != nil {
+			return fmt.Errorf("failed to get delete address set %s ops: %v", as.Name, err)
+		}
+		syncer.ops = append(syncer.ops, ops...)
+		return nil
+	}
+	staleAS := false
+	if ipSuffix == "" {
+		staleAS = true
+		if len(as.Addresses) > 0 {
+			ip := net.ParseIP(as.Addresses[0])
+			if ip != nil {
+				if ip.To4() != nil {
+					ipSuffix = ipv4AddressSetSuffix
+				} else {
+					ipSuffix = ipv6AddressSetSuffix
+				}
+			}
+		}
+	}
+	if ipSuffix == "" {
+		// couldn't extract ip family
+		klog.Infof("Found stale address set %s without ip family suffix and empty ips list, deleting", oldName)
+		ops, err := syncer.getDeleteASOps(acls, qoses, lrps, as)
+		if err != nil {
+			return fmt.Errorf("failed to get delete address set %s ops: %v", oldName, err)
+		}
+		syncer.ops = append(syncer.ops, ops...)
+		return nil
+	}
+	nbdbASv4, nbdbASv6 := syncer.asf.GetDbObjsForAS(dbIndex, nil)
+	var nbdbAS *nbdb.AddressSet
+	if ipSuffix == ipv4AddressSetSuffix {
+		nbdbAS = nbdbASv4
+	} else {
+		nbdbAS = nbdbASv6
+	}
+	// since we need to update addressSet.Name, which is an index and not listed in getNonZeroAddressSetMutableFields,
+	// we copy existing addressSet, update it as needed, and replace (delete and create) existing address set with the updated
```

Comment:
i thought there was some issue with trying to delete and recreate something with the same index in the same ovsdb transaction....

Reply:
or are the names actually different everytime? I thought they would be the same hash.

Reply:
Yeah the names will be different, because now the name includes all ExternalIDs: `dbIndex.OwnerController + "_" + dbIndex.OwnerObjectType + "_" + dbIndex.OwnerNamespace +
		"_" + dbIndex.OwnerName + "_" + dbIndex.InternalID`